src/lib/AutoControlledComponent.js

```diff
@@ -120,10 +131,24 @@ export default class AutoControlledComponent extends Component {
 
   componentWillReceiveProps(nextProps) {
     if (super.componentWillReceiveProps) super.componentWillReceiveProps(nextProps)
+    const { autoControlledProps } = this.constructor
 
-    // props always win, update state with all auto controlled prop
-    const newState = _.pick(nextProps, this.constructor.autoControlledProps)
-    if (!_.isEmpty(newState)) this.setState(newState)
+    // Solve the next state for autoControlledProps
+    const newState = _.transform(autoControlledProps, (acc, prop) => {
+      const isNextUndefined = _.isUndefined(nextProps[prop])
+      const propWasRemoved = _.has(this.props, prop) && isNextUndefined
+
+      // if next is defined then use its value
+      if (!isNextUndefined) acc[prop] = nextProps[prop]
+
+      // reinitialize state for props just removed / set undefined
+      else if (propWasRemoved) acc[prop] = getAutoControlledStateValue(nextProps, prop)
+
+      // otherwise, continue persisting the auto controlled state value
+      else acc[prop] = this.state[prop]
+    }, {})
+
+    this.setState(newState)
```

Comment:
The final `else` branch here deferred to copying the old state to persist it.  This meant new state was no longer ever empty.  This was unnecessary since setting state for `foo` does not remove state for `bar`.

I've removed the unnecessary state copying and added a check to only set state if newState is not empty.